streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -211,28 +209,45 @@ public synchronized void start() {
      * and then wait for them to join.
      *
      * @throws IllegalStateException if process has not started yet
+     * @param timeout   how long to wait for {@link StreamThread}s to shutdown
+     * @param timeUnit  unit of time used for timeout
      */
-    public synchronized void close() {
+    public synchronized void close(final long timeout, final TimeUnit timeUnit) {
         log.debug("Stopping Kafka Stream process");
 
-        if (state == RUNNING) {
-            // signal the threads to stop and wait
-            for (final StreamThread thread : threads)
-                thread.close();
-
-            for (final StreamThread thread : threads) {
-                try {
-                    thread.join();
-                } catch (final InterruptedException ex) {
-                    Thread.interrupted();
-                }
+        if (state.compareAndSet(StreamsState.running, StreamsState.stopping)) {
```

Comment:
Why do we need an atomic here? `close()` should be called single threaded only, right? And if I miss anything, we do we not need to use atomic to switch from "created" to "running" in `start()`.